api/pull.go

```diff
@@ -0,0 +1,380 @@
+	case *transactionsFilter:
+		return api.getTransactions(current, f.(*transactionsFilter))
+	case *logsFilter:
+		return api.getLogs(current, f.(*logsFilter))
+	}
+
+	return nil, nil
+}
+
+// filterExpiryChecker continuously monitors all the registered filters and
+// if any of them has been idle for too long i.e. no requests have been made
+// for a period defined as filterExpiry using the filter ID it will be removed.
+func (api *PullAPI) filterExpiryChecker() {
+	for _ = range time.Tick(api.config.FilterExpiry) {
+		for id, f := range api.filters {
+			if f.expired() {
+				api.UninstallFilter(id)
+			}
+		}
+	}
+}
+
+func (api *PullAPI) addFilter(f filter) rpc.ID {
+	api.mux.Lock()
+	defer api.mux.Unlock()
+
+	// we limit max active filters at any time to prevent abuse and OOM
+	const maxFilters = 10_000
+	if len(api.filters) > maxFilters {
+		return ""
+	}
+
+	api.filters[f.id()] = f
+	return f.id()
+}
+
+func (api *PullAPI) getBlocks(latestHeight uint64, filter *blocksFilter) ([]common.Hash, error) {
+	nextHeight := filter.nextHeight()
+	hashes := make([]common.Hash, latestHeight-nextHeight)
+
+	// todo we can optimize if needed by adding a getter method for range of blocks
+	for i := nextHeight; i <= latestHeight; i++ {
+		b, err := api.blocks.GetByHeight(i)
+		if err != nil {
+			return nil, err
+		}
+		h, err := b.Hash()
+		if err != nil {
+			continue
+		}
+		hashes[i] = h
+	}
+
+	return hashes, nil
+}
+
+func (api *PullAPI) getTransactions(latestHeight uint64, filter *transactionsFilter) (any, error) {
+	txs := make([]models.Transaction, 0)
+	hashes := make([]common.Hash, 0)
+
+	// todo we can optimize if needed by adding a getter method for range of txs by heights
+	for i := filter.nextHeight(); i <= latestHeight; i++ {
+		b, err := api.blocks.GetByHeight(i)
+		if err != nil {
+			if errors.Is(err, errs.ErrNotFound) { // block not found
+				fmt.Println("BLOCK NOT FOUND ", i)
+				return nil, nil
+			}
+			return nil, err
+		}
+
+		// for now there will only be one tx per block
+		for _, h := range b.TransactionHashes {
+			tx, err := api.transactions.Get(h)
+			if err != nil {
+				return nil, err
+			}
+
+			h, err := tx.Hash()
+			if err != nil {
+				continue
+			}
+			txs = append(txs, tx)
+			hashes = append(hashes, h)
+		}
+	}
+
+	if filter.fullTx {
+		return txs, nil
+	}
+
+	return hashes, nil
+}
+
+func (api *PullAPI) getLogs(latestHeight uint64, filter *logsFilter) (any, error) {
+	last := big.NewInt(int64(filter.nextHeight()))
+	latest := big.NewInt(int64(latestHeight))
+	criteria := logs.FilterCriteria{
+		Addresses: filter.criteria.Addresses,
+		Topics:    filter.criteria.Topics,
+	}
+
+	// if filter criteria "to" block is lower height than current height there can not be any new
+	// logs available, so we return nil, if "to" is negative it means latest, so we always return new data
+	to := filter.criteria.ToBlock
+	if to != nil && latest.Cmp(to) > 0 && to.Cmp(big.NewInt(0)) >= 0 {
+		return nil, nil
```

Comment:
Maybe we should return `[]*gethTypes.Log{}` instead of `nil`? I do not know what assumptions clients make, just mentioning it in any case.

Reply:
yes we should